File: Javascript/Algorithms/63.js
/**
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
var uniquePathsWithObstacles = function(obstacleGrid) {
  let m = obstacleGrid.length, n = obstacleGrid[0].length;
  if (obstacleGrid[0][0] || obstacleGrid[m - 1][n - 1]) return 0;

  let mat = [];
  for (let i = 0; i < m; i++) mat.push(Array(n).fill(0));
  mat[0][0] = 1;

  for (let r = 0; r < m; r++) {
    for (let c = 0; c < n; c++) {
      // if at obstacle, continue.
      if (obstacleGrid[r][c]) continue;

      // Add top/left if possible.
      if (r > 0 && obstacleGrid[r - 1][c] == 0) mat[r][c] = mat[r - 1][c];
      if (c > 0 && obstacleGrid[r][c - 1] == 0) mat[r][c] += mat[r][c - 1];
    }
  }
  return mat[m - 1][n - 1];
};

File: Python/Algorithms/63.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        # cases where we can't move from start/to end because it's blocked.
        if obstacleGrid[0][0] or obstacleGrid[-1][-1]:
            return 0

        # Build m x n matrix to hold paths.
        m, n = len(obstacleGrid), len(obstacleGrid[0])
        mat = [[0] * n for _ in range(m)]
        mat[0][0] = 1

        for r in range(m):
            for c in range(n):
                # If this path is blocked, move on.
                if obstacleGrid[r][c]:
                    continue

                # if not, check if we had an obstacle on top.
                if r and obstacleGrid[r-1][c] == 0:
                    mat[r][c] = mat[r-1][c]
                # or to our left.
                if c and obstacleGrid[r][c-1] == 0:
                    mat[r][c] += mat[r][c-1]

        # end will hold total number of paths.
        return mat[-1][-1]
